### PR TITLE
Replace EngineClosedException with AlreadyClosedExcpetion

### PR DESCRIPTION
**Pull request by bleskes**
`EngineClosedException` is a ES level exception that is used to indicate that the engine is closed when operation starts. It doesn't really add much value and we can use `AlreadyClosedException` from Lucene (which may already bubble if things go wrong during operations). Having two exception can just add confusion and lead to bugs, like wrong handling of `EngineClosedException` when dealing with document level failures. The latter was exposed by `IndexWithShadowReplicasIT`.

This PR also removes the AwaitFix from the `IndexWithShadowReplicasIT` tests (which was what cause this to be discovered). While debugging the source of the issue I found some mismatches in document uid management in the tests. The term that was passed to the engine didn't correspond to the uid in the parsed doc - those are fixed as well.

`EngineClosedException` is kept around for BWC. Once this goes into 5.3, I will remove it from master.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -544,6 +544,10 @@ final Exception checkIfDocumentFailureOrThrow(final Operation operation, final E
             // and set the error in operation.setFailure. In case of environment related errors, the failure
             // is bubbled up
             isDocumentFailure = maybeFailEngine(operation.operationType().getLowercase(), failure) == false;
+            if (failure instanceof EngineClosedException) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I am not sure what makes a `EngineClosedException` not a `isDocumetnFailure` but `AlreadyClosedException` does. I think we use the boolean returned from `maybeFailEngine` in the wrong way here. It almost seems like that we need some state handling on `Operation` that tells us if the document actually failed to index or if we didn't try or if we tried, succeeded but after success we failed due to concurrent closing. This entire exception based decision making here seems error prone to me.

**Review by s1monw — CHANGES_REQUESTED** [review]
I left one comment

**Comment by bleskes** [comment]
@s1monw as discussed, I update the PR to stop using EngineClosedException. Can you take a look? (PR title and description are adapted as well)

File: core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 2]
```diff
@@ -388,8 +386,6 @@ private UpdateResultHolder executeUpdateRequest(UpdateRequest updateRequest, Ind
                         Exception failure = operationResult.getFailure();
                         assert failure instanceof VersionConflictEngineException
                             || failure instanceof MapperParsingException
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: 👍

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 3]
```diff
@@ -431,12 +431,12 @@ public void testStalePrimaryShardOnReroute() throws InterruptedException {
...
     }
 
-    private ElasticsearchException randomRetryPrimaryException(ShardId shardId) {
+    private Exception randomRetryPrimaryException(ShardId shardId) {
         return randomFrom(
             new ShardNotFoundException(shardId),
             new IndexNotFoundException(shardId.getIndex()),
             new IndexShardClosedException(shardId),
-            new EngineClosedException(shardId),
+            new AlreadyClosedException("primary is closed"),
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: maybe put the shard ID into the message just for completeness?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -1667,7 +1666,7 @@ public void onFailedEngine(String reason, @Nullable Exception failure) {
     private Engine createNewEngine(EngineConfig config) {
         synchronized (mutex) {
             if (state == IndexShardState.CLOSED) {
-                throw new EngineClosedException(shardId);
+                throw new AlreadyClosedException("can't create engine - shard is closed");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: should we include the shard ID in the message in general?

**Review by s1monw — APPROVED** [review]
left some suggestions

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I'm partial. can do...

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java (hunk 3 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: this is a test only exception, but sure

**Comment by bleskes** [comment]
retest this please

**Comment by bleskes** [comment]
Thx @s1monw

**Pull request merged**